## zadatak 1

Napiši program koji učitava listu i briše sve duplikate iz te liste te ispisuje novu listu s 
obrisanim duplikatima.

In [1]:
li = [2,34,5,5,5,2,2,4,89,4,4,2,5,7,8,4,5,6,69,42,420,1729]

def remove_duplicates(e):
	return list(set(e))

remove_duplicates(li)

[1729, 2, 34, 4, 5, 6, 7, 8, 69, 42, 420, 89]

## zadatak 2

Napravi obrtaljku od unesenog stringa te na svakom drugom mjestu pridruzi nasumičan broj.

In [2]:
import random

string = "nekakav proizvoljni string"
string_reversed = string[::-1]

for i in string_reversed[::2]:
	print(f"{i} {str(random.randint(0,100))}")

g 11
i 43
t 87
  79
n 64
l 26
v 15
i 92
r 22
  66
a 60
a 20
e 46


## zadatak 3

Napiši program koji upisuje bodove n plesnih natjecanja. Ispiši zbroj svih bodova tako da odbaciš 
najbolji i najlošiji rezultat.

In [3]:
awarded_points = [2,3,7,8,6,3,1]

awarded_points.sort()

sum_pts = 0
for i in awarded_points[1:-1]:
	sum_pts += i

print(sum_pts)

21


## zadatak 4

Kreiraj listu brojeva i pretvori svaki element u listi u njegov kvadrat

In [4]:
rand_nums = [2,3,7,8,6,3,1]

def get_square(e):
	return [i**2 for i in e]

get_square(rand_nums)

[4, 9, 49, 64, 36, 9, 1]

## zadatak 5

Kreiraj listu koja se sastoji od stringova i brojeva, te odvoji brojeve i stringove u zasebne liste

In [5]:
str_and_num_li = ['Python', 69, 'C++', 1729, 'JavaScript', 420, 'nesto']

def get_ints(e):
	return [i for i in e if isinstance(i, int)]

def get_strs(e):
	return [i for i in e if isinstance(i, str)]

print(get_ints(str_and_num_li))
print(get_strs(str_and_num_li))

[69, 1729, 420]
['Python', 'C++', 'JavaScript', 'nesto']


## zadatak 6

S tipkovnice učitavajte cijele brojeve. Prvi upisani broj može biti bilo 
koji cijeli broj. Učitavanje ponavljati dok god je upisani broj strogo 
veći od prethodno upisanog broja. Ispisati sumu svih učitanih brojeva 
osim broja zbog kojeg je prekinuto učitavanje.

In [10]:
user_list = []

def user_input():
	while True:
		inp = int(input("Enter a word: "))
		user_list.append(inp)
		if user_list[len(user_list) - 1] < user_list[len(user_list) - 2]:
			break
	total = 0
	for i in range(len(user_list) - 1):
		total += user_list[i]
	return total, user_list

user_input()

(20, [5, 6, 9, 5])

## zadatak 7

Učitajte s tipkovnice 2 niza znakova, svaki od tih nizova znakova 
spremite u zasebnu varijablu. Ispišite indekse na kojima se pojavljuju 
ista slova neovisno o veličini ('a' i 'A' tretirati jednako).

In [16]:
first_srt = "tjhreWEt"
second_str = "dfHJflhdsj"

for i, (first_idx, second_idx) in enumerate(zip(first_srt, second_str)):
	if first_idx.lower() == second_idx.lower():
		print(i)

2


## zadatak 8

Napišite program koji s tipkovnice učitava proizvoljni cijeli 
troznamenkasti broj. Ako učitani broj nije troznamenkasti, ispišite 
poruku o greški i prekinite daljnje izvođenje programa. U slučaju da 
je učitani broj ispravan, ispišite prvi sljedeći troznamenkasti 
palindrom. Na primjer, ako je učitani broj 120, prvi sljedeći palindrom 
je 121.

In [3]:
def is_palindrome(num):
	return str(num) == str(num)[::-1]

def closest_palindrome(num):
	if len(str(num)) == 3:
		next_palindrome = num + 1
		while not is_palindrome(next_palindrome):
			next_palindrome += 1
		return next_palindrome
	else:
		print(f"{num} is invalid number for given condition!")

closest_palindrome(int(input("Enter a 3 digit number: ")))

232

## zadatak 9

Napišite program koji učitava cijele brojeve sve dok je unesena 
vrijednost veća od 0. Pronađite koji od učitanih brojeva ima najmanju 
sumu znamenki te ispišite taj broj i sumu.

In [1]:
def get_digit_sum(e):
	sum = 0
	for digit in str(e): 
		sum += int(digit)
	return sum

def list_sum():
	sum_list = []
	for i in li:
		sum_list.append(get_digit_sum(i))
	return sum_list

li = []
while True:
	inp = int(input("Enter a number bigger than 0: "))
	if inp < 0:
		break
	else:
		li.append(inp)

print(min(li))
print(min(list_sum()))

22
4


## zadatak 10

S tipkovnice učitajte proizvoljni niz znakova. Kreirajte novi niz 
znakova koji će sadržavati naizmjence velika i mala slova iz ulaznog 
niza, redom kako se pojavljuju u ulaznom nizu: prvo veliko slovo u 
ulaznom nizu, prvo sljedeće malo slovo u nastavku ulaznog niza, 
prvo sljedeće veliko slovo u nastavku ulaznog niza itd. Novokreirani 
niz ispišite na zaslon. U nastavku se nalazi primjer:

Ulazni niz: ifeFemFEkej83FkW
Izlazni niz: FeFkFkW

In [12]:
input_string = "ifeFemFEkej83FkW"
output_string = ""

upper_case = True

for i in input_string:
	if "A" < i < "Z" and upper_case:
		output_string += i
		upper_case = False
	elif "a" < i < "z" and not upper_case:
		output_string += i
		upper_case = True

print(output_string)

FeFkFkW


## zadatak 11

Napišite program koji s tipkovnice učitava cijeli broj n iz intervala [3, 
20]. U slučaju da je unesena vrijednost neispravna, ispisati prikladnu 
poruku na ekran te zatražiti ponovni unos cijelog broja. Nakon 
učitane vrijednosti n, učitajte n parova cijelih brojeva. Nakon što je n
parova brojeva učitano, ispišite parove brojeva koji imaju najveću 
sumu.

**nije mi jasno što se tu traži**

## zadatak 12

Napišite program koji s tipkovnice učitava proizvoljni niz znakova. 
Nad učitanim nizom znakova napravite analizu je li taj niz palindrom 
ili nije. Niz je palindrom ako se isto čita slijeva nadesno ili pak 
zdesna nalijevo.

In [26]:
inp = input("Enter something: ")

def is_palindrome(e):
	return e == e[::-1]

is_palindrome(inp)

True

## zadatak 13

Napišite program koji će inicijalizirati varijablu n na proizvoljnu 
cjelobrojnu vrijednost. Vrijednost varijable n neka predstavlja red 
tablice. Ispisati tablicu veličine n redaka i n stupaca. Vrijednost 1 
neka se nalazi na glavnoj dijagonali, a vrijednost 0 na svim ostalim 
mjestima. U nastavku slijedi primjer za n=5:

```
1 0 0 0 0
0 1 0 0 0
0 0 1 0 0
0 0 0 1 0
0 0 0 0 1
```

Napomena: za ispis vrijednost, tako da se nakon ispisa vrijednosti 
dane print() funkciji ne ispiše novi redak koristiti sljedeću 
sintaksu: print(vrijednost, end='')


In [60]:
def draw_board(n):
	for i in range(n):
		for j in range(n):
			if i == j:
				print("1", end = " ")
			else:
				print("0", end = " ")
		print()

draw_board(5)

1 0 0 0 0 
0 1 0 0 0 
0 0 1 0 0 
0 0 0 1 0 
0 0 0 0 1 


## zadatak 14

Odaberite proizvoljno koordinatu T=(x,y), vrijednosti varijabli x
(stupac) i y (redak) neka budu manje od 10. Program neka ispiše 
polje 10x10 čiji su svi elementi vrijednosti "-" osim koordinate T čija 
je vrijednost "X". 

```

Primjer: T=(1,1)
- - - - - - - - - -
- X - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
- - - - - - - - - -
```

In [10]:
def print_coordinates(x, y):
	for i in range(10):
		for j in range(10):
			if i == x and j == y:
				print("x", end = " ")
			else:
				print("-", end = " ")
		print()

print_coordinates(5, 5)

- - - - - - - - - - 
- - - - - - - - - - 
- - - - - - - - - - 
- - - - - - - - - - 
- - - - - - - - - - 
- - - - - x - - - - 
- - - - - - - - - - 
- - - - - - - - - - 
- - - - - - - - - - 
- - - - - - - - - - 


## zadatak 15

Napišite program koji ispisuje koliko ima prostih brojeva između 
dva proizvoljna broja (prost broj je broj koji je djeljiv samo sa 1 i sa samim sobom).

In [164]:
def is_prime(x):
	if x < 2:
		return False
	for i in range(2, x):
		if x % i == 0:
			return False
	return True

def prime_nums(min, max):
	return [i for i in range(min, max + 1) if is_prime(i)]

prime_nums(2, 50)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]

## zadatak 16

Napišite program koji sadrži varijablu u kojoj je upisan proizvoljni 
niz znakova i brojčanu varijablu n. Provjerite je li vrijednost varijable 
n manja od broja znakova u nizu. Ako je vrijednost varijable n veća 
ispišite informaciju o grešci. Ispišite iz niza znakova svako n-to 
slovo. Na primjer, ulazni niz je "ABCDEFGH", n je 2, tada je izlaz 
"ACEG".

In [27]:
n = 2
string = "ABCDEFGH"

def string_checker(string, n):
	li = []
	if n > len(string):
		print("Error. N is bigger than string.")
	else:
		for i, j in enumerate(string):
			if i % n == 0:
				li.append(j)
	return li

print(string_checker(string, n))

['A', 'C', 'E', 'G']


## zadatak 17

Napišite program koji sadrži varijablu u kojoj je upisan proizvoljni niz 
znakova. Ispišite koliko velikih slova se nalazi u nizu. Ako je neko od 
unesenih slova u nizu "A", brojanje velikih slova je potrebno prekinuti 
i ispisati informaciju da je veliko slovo "A" pronađeno.


In [31]:
string = "fhVfGtDsdAfjQWongS"
counter = 0

for i in string:
    if i == 'A':
        print("Letter A found.")
        break
    if i.isupper():
        counter += 1

print(counter)

Letter A found.
3


## zadatak 18

Ispišite sve parne brojeve između 1 i 1000 koju su istovremeno 
djeljivi i s 5 i s 13.

In [38]:
def divisable_nums():
	li = [i for i in range(1, 1001) if i % 5 == 0 and i % 13 == 0]
	for j in li:
		if j % 2 == 0:
			print(j)

divisable_nums()

130
260
390
520
650
780
910


## zadatak 19

Napisi program koji unosi n brojeva i od znamenke desetice svakog broja stvara novi broj.

In [42]:
generated_number = ""

while True:
    inp = input("enter number, 0 to quit: ")
    if inp == "0":
        break
    second_digit = inp[-2] if len(inp) >= 2 else 0
    generated_number += second_digit

print(generated_number)

22


## zadatak 20

Napisi program koji unosi n brojeva i sastavlja novi broj od najvece znamenke u svakom od
brojeva

In [43]:
generated_number = ""

while True:
    inp = input("enter number, 0 to quit: ")
    if inp == "0":
        break
    biggest_digit = max(inp)
    generated_number += biggest_digit

print(generated_number)

33


## zadatak 21

Ispišite sljedeće primjere koristeći funkcije i petlju. funkcija mora primati varijabilan broj redaka koji ispisuje (npr. pod a), unesen broj je 7, jer ima 7 redaka. pod b) je unesen broj 5, jer ima 5 redaka)

```
a)

*
**
***
****
***
**
*

b)

  *  
 *** 
*****
 *** 
  * 
c)
1010101
 10101 
  101  
   1  
d) 
x x x x
x x x
x x
x
x x
x x x
x x x x
```

In [28]:
def draw_pattern_a(n):
	for i in range (0,n):
		for j in range (i):
			print("*", end="")
		print()

	for i in range (n,0,-1):
		for j in range (i):
			print("*", end="")
		print()

def draw_pattern_b(n):
	for i in range(1, (n+1)//2 + 1):
		for j in range((n+1)//2 - i):
			print(" ", end = "")
		for k in range((i*2)-1):
			print("*", end = "")
		print()

	for i in range((n+1)//2 + 1, n + 1): 
		for j in range(i - (n+1)//2):
			print(" ", end = "")
		for k in range((n+1 - i)*2 - 1):
			print("*", end = "")
		print()

def draw_pattern_c(n):
	for i in range((n+1)//2 + 1, n + 1): 
		for j in range(i - (n+1)//2):
			print(" ", end = "")
		for k in range((n+1 - i)*2 - 1):
			if k % 2 == 0:
				print("1", end = "")
			else:
				print("0", end = "")
		print()

def draw_pattern_d(n):
    if n%2 == 0:
        for i in range(n//2, 0, -1):
            print('x ' * i)
        for i in range(1, n//2 + 1):
            print('x ' * i)

draw_pattern_a(4)
draw_pattern_b(5)
draw_pattern_c(8)
draw_pattern_d(8)


*
**
***
****
***
**
*
  *
 ***
*****
 ***
  *
 1010101
  10101
   101
    1
x x x x 
x x x 
x x 
x 
x 
x x 
x x x 
x x x x 


## zadatak 22

Napišite program za koji traži najveći zajednički djelitelj dvaju brojeva

In [47]:
def gcd(a, b):
	while b:
		a, b = b, a % b
	return a

gcd(12, 18)

6

## zadatak 23

koristenjem funkcija i naucenog, napisi program koji racuna iznos racuna u ovisnosti o transakcijama. dnevnik transakcija prikazi ovako (D: 100, W: 200), dok ukupan iznos na racunu prikazuje sa (T: 1000). ako korisnik unese T = 1000 te povuce iznos od 350 (W: 350), aplikacija treba ispisati azurirano stanje T: 650. ako nakon toga korisnik uplati iznos na racun, npr. 200 (D: 200), program treba uvecati iznos racuna i azurirati ga (T: 850). unosom stringa "quit", program prekida se izvodjenje programa.

In [5]:
total = 0

def withdraw(total, w_amount):
    total -= w_amount
    return total

def deposit(total, d_amount):
    total += d_amount 
    return total

while True:
    inp = input("d for deposit, w for withdrawal, q to quit: ")

    if inp == "w":
        w_amount = int(input("Enter withdrawal amount: "))
        total = withdraw(total, w_amount)
        print(f"Total: {total}")

    elif inp == "d":
        d_amount = int(input("Enter deposit amount: "))
        total = deposit(total, d_amount)
        print(f"Total: {total}")

    elif inp == "q":
        break

Total: 500
Total: 300


## zadatak 24

zadan je rjecnik:
```
automobil = {
    'kilometri' : 230000,
    'oprema' : ['zimske gume', 'zracni jastuk', 'protuprovalni bacac plamena'],
}
```
1. kilometre preuredite tako da dodatno broji i kilometre do servisa. postavite vrijednost na 300.
2. iz opreme, promijenite "zimske gume" u "ljetne gume" te "zracni jastuk" u "zracni balon"
3. dodajte polje za unos specifikacije koja sadrzi godinu proizvodnje (1976.), potrosnju (15 litara) i max. brzinu (178 km/h)
4. dodajte cijenu i postavite je na 45000
5. dodajte osiguranje (npr. croatia) te postavite vrijednost na True
6. kada se pozove funkcija 'servis', znaci da se automobil servisirao i promijenile su se sljedece vrijednosti:

- kilometri do servisa su se postavili na 2000
- zracni balon postao je 'novi zracni jastuk'
- ljetne gume promijenjene su na 'cjelogodisnje gume'
- cijenu uvecajte za 2000

In [6]:
automobil = {
    'kilometri' : 230000,
    'oprema' : ['zimske gume', 'zracni jastuk', 'protuprovalni bacac plamena'],
}

automobil['kilometri'] = {'kilometri': automobil['kilometri'], 'kilometri do servisa': 300}

automobil["oprema"][0] = "ljetne gume"
automobil["oprema"][1] = "zracni balon"

automobil["specifikacija"] = [1976, 15, 178]

automobil["cijena"] = 45000

automobil['osiguranje'] = {'croatia': True}

def servis():
    automobil['kilometri']['kilometri do servisa'] = 2000
    automobil['oprema'][0] = 'cjelogodisnje gume'
    automobil['oprema'][1] = 'novi zracni jastuk'
    automobil['cijena'] += 2000

servis()

# 2do za vikend - nisam stigao do roka

## zadatak 25

napisite funkciju koja prima string oblika "\<predmet\>, \<cijena\>, \<kolicina\>" (npr. "kruh, 7.5, 5"), koji bi simulirao policu neke trgovine. program treba razdvojiti navedeni string te zasebne elemente spremiti u rjecnik. rjecnik se treba spremiti u listu proizvoda. simuliraj kupovinu: kada kupac kupi proizvod (ili broj proizvoda), njegova kolicina se treba umanjiti za iznos. ukoliko je kolicina 0, predmet se treba ukloniti sa police.

## zadatak 26

vas omiljeni carobni aparat za kavu je podivljao. aparat treba ispitivati korisnika za kavu sve dok on ne unese "Q" ili 0. aparat prihvaca samo kovanice od 2kn ili 5kn. kada korisnik ubaci kovanicu od 5kn, aparat nausmicno izbacuje dvije kave. kada korisnik ubaci kovanicu od 2kn, aparat nasumicno izbaci jednu kavu. ako aparat sakupi 20kn, on se magicno napuni za 3 (nasumicne) kave. neka set sljedeceg oblika simbolizira pocetnu kolicinu kave u aparatu: dict_kava = {"turska": 4, "latte": 3, "espresso": 7, "makiyato": 5}. 

## zadatak 27

nadogradite prethodni zadatak na nacin da ubacite u aparat i opciju za dodavanje mlijeka, secera ili meda u kavu. kada korisnik ubaci kovanicu, aparat ga treba pitati zeli li kavu sa dodatkom ili ne. postavite vrijednost svakog dodatka na 5, koristeci fromkeys. ukoliko korisnik zatrazi kavu sa sladilom kojega vise nema u aparatu, treba ispisati odgovarajucu poruku te ponuditi korisniku opciju bez dodatka ili povrat novca, nakon cega zavrsava sa izvodjenjem. u svim slucajevima treba se ispisati kakva je kava isporucena, sa kojim sladilom, koliko novca je ubaceno te koliki je iznos novca u aparatu nakon isporuke kave.

## zadatak 28

Neka program omoguci unos do trenutka kad je unesen broj -1. Korisnik unosi stringove, a program vraca slucajne recenice iz zbirke mogucih odgovora.

## zadatak 29

napravite program koji ce simulirati izvlacenje karti iz spila. program mora imati sljedece:

- spil od 20 karti
- 4 razlicite boje
- 5 brojeva za svaku boju

program treba imati mogucnost povlacenja 3 karte, povlacenja 1 karte, mijesanja spila, zbrajanja bodova (brojevi na kartama).
kreirajte izbornik sa mogucnoscu izbora opcije (zapocni igru, zavrsi igru) -> ako korisnik odabere "zapocni igru", spil se mijesa te se povlace 3 karte. igraca se zatim pita hoce li povuci dodatnu kartu ili ne. ako odabere novu kartu, ona se dodaje u ruku; u suprotnom se broj bodova usporedjuje sa protivnikom. koristeci random varijablu (12-20), kreirajte protivnika, gdje random varijabla predstavlja njegov broj bodova. program treba racunati trenutni broj bodova te ako on prelazi 20, ispisati "izgubio si" i pitati korisnika zeli li ponovno igrati ili ne. ukoliko igrac pobijedi, treba ispisati poruku: pobijedio si i postaviti pitanje za novom igrom

## zadatak 70

kreirajte password generator, koji mora sadrzavati sljedece funkcionalnosti:

1. prilikom generiranja treba prikupiti sljedece podatke od korisnika:

- koliko lozinki zeli napraviti
- da li da koristi slova, brojeve, simbole
- duljinu lozinke
- kapitalizaciju

2. nadogradite program tako da korisnik upise rijec (hint). od hint se treba ubaciti u nasumicnom redosljedu i kapitalizaciji u lozinku koja se generira

## zadatak 71

napravite tipicnog "chat bota" neke drzavne firme. bot treba primati input korisnika te ako se odredjena rijec u stringu poklapa sa nekim od njemu predefiniranih stringova ili rijeci, ispisati odgovarajucu poruku. npr. ako je korisnik unio "pozdrav, ja sam Alen i radim za mirovinsko osiguranje". bot treba kao prvi string ispisati: "pozdrav, ja sam bot i napravljen pomocu AI", zatim (zbog rijeci mirovinsko), ispisati: " hrvatski zavod za mirovinsko osiguranje je trenutno na pauzi". unesite po zelji 20ak razlicitih stringova te simulirajte neki razgovor/temu po zelji.